worker/codalabworker/local_run/docker_image_manager.py

```diff
@@ -3,218 +3,194 @@
-                time.sleep(self._sleep_secs)
+        if self._max_image_cache_size:
 
-        self._main_thread = threading.Thread(target=loop, args=[self])
-        self._main_thread.start()
+            def cleanup_loop(self):
+                while not self._stop:
+                    try:
+                        self._cleanup()
+                        self._save_state()
+                    except Exception:
+                        traceback.print_exc()
+                    time.sleep(self._sleep_secs)
+
+            self._cleanup_thread = threading.Thread(target=cleanup_loop, args=[self])
+            self._cleanup_thread.start()
 
     def stop(self):
         logger.info("Stopping docker image manager")
         self._stop = True
         self._downloading.stop()
-        self._main_thread.join()
-        logger.info("Stopped docker image manager. Exiting. Exiting.")
+        self._cleanup_thread.join()
+        logger.info("Stopped docker image manager.")
 
-    def _process_images(self):
+    def _get_image_disk_usage(self):
         """
-        Transition image states. Also remove FAILED states that are too old.
+        Fetches Virtual and Marginal disk sizes of all the docker images that this image manager
+        has in its last used tracking state.
+        :returns: List[Tuple[int, str, int, int]]: Sorted list of (last used timestamp, image id, virtual size, marginal size)
         """
-        now = time.time()
-        with self._lock:
-            for digest, state in self._images.items():
-                if state.stage == DependencyStage.FAILED:
-                    if now - state.last_used > self._max_age_failed_seconds:
-                        del self._images[digest]
-
-            for digest, state in self._images.items():
-                self._images[digest] = self.transition(state)
+        image_disk_use = []
+        for digest, last_used in self._last_used.items():
+            try:
+                image_info = self._docker.images.get(digest)
+                virtual_size, marginal_size = (
+                    image_info.attrs['VirtualSize'],
+                    image_info.attrs['Size'],
+                )
+                image_disk_use.append(last_used, image_info.id, virtual_size, marginal_size)
+            except docker.errors.ImageNotFound:
+                continue
+        return sorted(image_disk_use)
 
     def _cleanup(self):
         """
-        If Docker's disk usage is higher than the quota given at initialization, clean up old images
-        until disk usage is under the quota again. Remove images in this order:
-            - FAILED images, least recently touched first
-            - READY images, least recently touched first
-        For now, if all the images are DOWNLOADING, log but don't delete. Images will be deleted after
-        download completes
-
+        Prunes the image cache for runs
+        1. Only care about images we downloaded and know about
```

Comment:
'we downloaded' means actually CodaLab downloaded or anyone downloaded?

Reply:
CodaLab downloaded, i.e. anything that we're keeping track of in the 'last used' dict.